flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/nodes/dataset/DataSetUnion.scala

```diff
@@ -69,7 +73,7 @@ class DataSetUnion(
       rows + metadata.getRowCount(child)
     }
 
-    planner.getCostFactory.makeCost(rowCnt, 0, 0)
+    planner.getCostFactory.makeCost(if (all) rowCnt else rowCnt * 0.1, 0, 0)
```

Comment:
The cost for union should be higher than for union all. Also, `rowCnt` is the number of rows processed, not the result size.
How about this?
```
planner.getCostFactory.makeCost(
  rowCnt, 
  if (all) 0 else, rowCnt,
  if (all) 0 else rowCnt)
```